pkg/operator/ceph/cluster/cluster.go

```diff
@@ -304,7 +350,7 @@ func clusterChanged(oldCluster, newCluster cephv1.ClusterSpec, clusterRef *clust
 
 	// any change in the crd will trigger an orchestration
 	if !reflect.DeepEqual(oldCluster, newCluster) {
```

Comment:
I think I'm finally able to express the concerns I have here a little better. The check for difference is first done with `reflect.DeepEqual`, and the diff is generated with `cmp.Diff`. But this is like asking 2 different people if two versions of something are identical: those 2 people may have 2 slightly different ideas about equality, and we can't truthfully report a diff written by someone who didn't decide equality in the first place.

The github issue you found and copied the `deepAllowUnexported` function from is interesting, and I think it exposes that our current use of `reflect.DeepEqual` is flawed: if k8s changes its internal representations of something, we may have a broken check, and this is in our application code, not just in a test.

It seems that the `cmp` package is superior, and we should use something from `cmp` instead of `reflect` for the if block beforehand as well. It also seems that we have no choice but to implement the custom comparator for `Quantity` as you had before since that was defined by the public API. Unfortunately, this means that we must implement a compartor for all custom types Kubernetes implements in the future for resources we reuse, but it's my hope that Quantity will be all we need to implement. It does seem better to break on something in the public API which we failed to update rather than failing on something that is implementation-specific.

We also now know that `Diff` will output a panic if we haven't implemented something, and we should never panic during runtime if we can help it. In addition to implementing both the check and diff with `cmp`, we should wrap the `if` block in a panic catcher and log an error if the compare or diff fails.

We are heavily reliant on the `if` check being accurate in Rook, and I feel like we should have some other opinions about the best way to proceed. If the `if` check using `cmp` fails to determine equality and panics, should we just assume that the cluster has changed, or could that cause too many updates?  Can we fall back to a check with `AllowUnexported`? Can we fall back to `reflect.DeepEqual`? Is there a different, better fallback option? Is there a different/better way to do this entirely?

@travisn do you have thoughts?

Reply:
Printing the diff will help someone troubleshoot a cluster when they look at the logs. Seems like we should do the following:
- Add a comparer for each place where we would get a known exception, for example with the quantity.
- Ensure the operator doesn't crash, so we need to recover from a panic raised by `cmp.Diff()`. In that case, log that the CR changed in an unknown way.

Reply:
is the suggestion to use recover() ?

Reply:
if we were to implement defer+recover, should we keep the cluster as is (treat it as though nothing has changed) and log?

Reply:
`reflect.DeepEqual()` may not be perfect, but it seems sufficiently accurate. If `reflect.DeepEqual()` says something changed, I would vote to go ahead and trigger the orchestration. It's ok to trigger an extra orchestration. The reason we added this trigger is that the orchestration wasn't happening often enough.